src/hotspot/share/opto/phaseloadfolding.cpp

```diff
@@ -0,0 +1,422 @@
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ *
+ */
+
+#include "libadt/vectset.hpp"
+#include "memory/resourceArea.hpp"
+#include "opto/addnode.hpp"
+#include "opto/callnode.hpp"
+#include "opto/cfgnode.hpp"
+#include "opto/compile.hpp"
+#include "opto/memnode.hpp"
+#include "opto/mulnode.hpp"
+#include "opto/node.hpp"
+#include "opto/phaseloadfolding.hpp"
+#include "utilities/globalDefinitions.hpp"
+
+void PhaseLoadFolding::optimize() {
+  ciEnv* env = C->env();
+  if (env->should_retain_local_variables() || env->jvmti_can_walk_any_space()) {
+    // Give up because JVMTI can do wonders
+    return;
+  }
+
+  // This phase is very fast, but it is still preferable not to allow potential unbounded
+  // iterations
+  for (int i = 0; i < 10; i++) {
+    bool progress = do_optimize();
+    if (!progress) {
+      return;
+    }
+    _igvn.optimize();
+    if (C->failing()) {
+      return;
+    }
+  }
+}
+
+// The escape status of a node is visible in the memory graph. That is, at runtime, if a load 'l'
+// from an object 'o' must be executed after an action 'a' that allows 'o' to escape, and in the
+// IR graph, the node 'L' corresponding to 'l' consumes the address 'O' + c, with 'O' being the
+// node corresponding to the newly allocated object 'o', then there must be a path along the
+// use-def edges from 'L' to the node 'A' that corresponds to 'a'.
+//
+// - If 'a' is a method invocation that receives 'o' as an argument, then in the graph, 'A' kills
+//   all memory. As a result, the memory input of 'L' must be a transitive use of 'A'. This is
+//   because in a well-behave memory graph, there is always a path of use-def edges from a memory
+//   node to the previous bottom memory node. This is important as it ensures memory fences can
+//   serialize memory operations by imposing use-def dependencies between the fence and the
+//   surrounding memory nodes.
+//   Example:
+//       Integer o = new Integer(v);
+//       int x = o.value;
+//       if (flag) {
+//         consume(o);
+//         int y = o.value;
+//       } else {
+//         consume(null);
+//         int z = o.value;
+//       }
+//       int t = o.value;
+//   The memory graph will then look like:
+//        NarrowMemProj (Integer.value)
+//          |          |
+//          |          |
+//     CallJava(o) CallJava(null)
+//          |          |
+//          |          |
+//        Proj1      Proj2
+//           \       /
+//            \     /
+//              Phi
+//   We can see that the object can be considered non-escape at NarrowMemProj, CallJava(null), and
+//   Proj2, while it is considered escape at CallJava(o), Proj1, Phi. The loads x and z will be
+//   from NarrowMemProj and Proj2, respectively, which means they can be considered loads from an
```

Comment:
So this optimization is based on JDK-8327963 changes which introduced NarrowMemProj. But I don't see you can for it in code.